openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,198 @@
+            'username' has permissions to modify account information.
+        preference_key (string): The key for the user preference.
+        preference_value (string): The value to be stored.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        AccountUserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        AccountValidationError: the update was not attempted because validation errors were found with
+            the supplied update
+        PreferenceUpdateError: the update could not be completed.
+    """
+    existing_user = _get_user(requesting_user, username)
+    if preference_value is None or preference_value == '':
+        message = _('Preference {preference_key} cannot be set to an empty value').format(
+            preference_key=preference_key
+        )
+        raise PreferenceUpdateError(message, user_message=message)
+    try:
+        UserPreference.set_preference(existing_user, preference_key, preference_value)
+    except PreferenceNotFound:
+        raise PreferenceNotFound()
+    except Exception as error:
+        raise PreferenceUpdateError(
+            "Error thrown when setting preference: '{}'".format(error.message)
+        )
+
+
+def delete_user_preference(requesting_user, preference_key, username=None):
+    """Deletes a user preference on behalf of a requesting user.
+
+    Note:
+        It is up to the caller of this method to enforce the contract that this method is only called
+        with the user who made the request.
+
+    Arguments:
+        requesting_user (User): The user requesting to delete the preference. Only the user with username
+            'username' has permissions to delete their own preference.
+        preference_key (string): The key for the user preference.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        AccountUserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        PreferenceNotFound: the user does not have a preference with the specified key.
+        PreferenceUpdateError: the delete could not be completed.
+    """
+    existing_user = _get_user(requesting_user, username)
+    try:
+        UserPreference.delete_preference(existing_user, preference_key)
+    except PreferenceNotFound:
```

Comment:
Is it really necessary to catch and re-raise the same exception? I would just let it raise "naturally", and that way the error message (if there is one) is preserved.